### ISSUE TITLE
Add support for updating.

### ISSUE THREAD

**Issue by rofinn** [invenia/PkgTemplates.jl#16]
It isn't uncommon for a plugin to get updated, so it would be nice if we could run `PkgTemplates.update("MyPkg", ...)` to update the generated files in a package. Some of us might even want to automate this with a bot that opens PRs.

**Comment by rofinn** [comment]
I can see that there is a `force` option on `generate`, but it's unclear how well this will work and we should probably have specific logic for updating a repo.

**Comment by christopher-dG** [comment]
`force` deletes and recreates the repo brand-new, so you definitely don't want to do that. I think this is a good idea!

**Comment by oxinabox** [comment]
I have been thinking of this as:
`fixup(path, template)` that would take a template and would attempt to bring the package inline with it.

Also I think one way to make this simpler,
is that it should refuse to work except on a git repo with no unstaged/uncommitted changes.
So that afterwards one can do `git diff`,
and `git checkout -- fileididnotwantchanged`

**Comment by christopher-dG** [comment]
Yeah, that seems reasonable.

**Comment by oxinabox** [comment]
BBTW solving this issue: 
would let us do conversion of nonpackages to package.
Since it would fix the files to follow the template.
Which is a template for a package.

**Comment by tkf** [comment]
How about

1. create a temporary orphan branch
2. create a git worktree of the temporary branch in a temporary directory
3. expand the template in it
4. remove the worktree
5. try merging the temporary branch

?

This should work without a merge conflict if the template only contains plugins that were not used (e.g., adding GitHub Actions).  For updating the plugins, the user can use the usual merge conflict resolution workflow.

**Comment by oxinabox** [comment]
that seems reasonable yes.

**Comment by christopher-dG** [comment]
Some random thoughts on this:

I'd kind of like to not rely on Git. Obviously > 90% of packages generated with PkgTemplates are also Git repos but it's not a guarantee. I think maybe we could just back up the current state to a temp directory, print out a message saying "hey we're about to potentially overwrite a bunch of files, they're all backed up here, do you want to proceed?", and then if yes just write the files ignoring any current contents. Git users can then look at their diffs to see what changed. Maybe there could also be some check for "if it's already a Git repo, require that it's clean".

I think we would also need some kind of trait that determines update behaviour. For example, most people are not going to want their `src/mod.jl` overwritten by the `SrcDir` plugin.
Or maybe just a predicate:

```jl
# Never overwrite module file
should_update(::SrcDir, ::Template, pkg) = false
# Always overwrite single files
should_update(::FilePlugin, ::Template, pkg) = true
# Only generate docs if there are none
should_update(::Documenter, t::Template, pkg) = isdir(joinpath(t.dir, pkg, "docs"))
# ...
```

**Comment by tkf** [comment]
> Maybe there could also be some check for "if it's already a Git repo, require that it's clean".

It's possible that some files are ignored by Git and the repository may have Git submodules. So, checking if it is safe to overwrite some files is somewhat complicated. Other VCS may have similar gotchas, too. I think it'd be much safer to let the VCS handle these difficulties and avoid directly writing to user's work tree.

**Comment by DilumAluthge** [comment]
@christopher-dG Any further thoughts on this?

**Comment by christopher-dG** [comment]
Not really. If anyone else wants to take the lead on this, feel free -- I have a lot of stuff on my plate lately.
